More engine abstraction, work toward Peekable and Iterable for Snapshot

brson opened the pull request:
###  What have you changed?

This builds on https://github.com/tikv/tikv/pull/5835. The first new commit is "Add extension methods for creating RocksEngine references".

This includes a variety of work towards engine abstraction. Much of it is shaping the engine_traits abstractions in ways that will make the porting of the engine crate easier.

I was hoping to get a little further before submitting, but it's getting toward the end of the week, and the branch is getting large.

Some things this PR does:

- Documents what I have learned about the refactoring strategy we should use in `engine_traits/src/lib.rs`. This will be the basis of a call for participation.

- In `engine_rocks` add a `Compat` trait that contains a single method `c()`, implemented to convert from `&Arc<DB>` to `&RocksEngine`, and from `&Snapshot` to `&RocksSnapshot`, and `&SyncSnapshot` to `&RocksSyncSnapshot` etc. These are still not used, but will soon be sprinkled all over the tikv codebase as `engine` is merged into `engine_rocks`. That is why the method name is so short - to avoid syntactic noise. I have used these extensively in the branches I've discarded while producing this PR.

- Remove the `IterOptions` type from `engine` and instead reexport it from `engine_traits`. This is a useful migration strategy for "plain old data" types. `engine` now depends on `engine_traits`.

- Make several changes to `engine_traits` APIs to more closely match the APIs they are replacing, and for efficiency. This is again to make porting easier. Rationale for each is in the individual commit messages.

- Migrate callers of `Snapshot::db_iterator` to `engine_rocks`. Delete `db_iterator` and `db_iterator_cf`.

- Add a `DBVector` type to `engine_traits` etc. for data returned from database queries. Use it in the `Peekable` APIs. I strongly suspect using `Vec` for this was going to be an unacceptable deoptimization since it forces an extra memcpy on every query.

- Introduce `tikv::into_other` for converting between error types that don't have mutual dependencies. Implement `IntoOther<raft::Error> for engine_traits::Error` and `IntoOther<kvproto::Error> for engine_traits::Error`. Like `Compat` I have used these in the discarded branches I've created while building this PR and expect them to be important in the branches that end up working.

I started out this work attempting to migrate `Iterable for Snapshot` callers to engine_rocks, and ended up making a bunch of these preparatory commits; then attempting to migrate `Peekable for Snapshot` callers and doing the same.

Right now I think `Peekable` is a good target for the next steps, and my next goal is to convert `impl Peekable for RegionSnapshot` to use the `Peekable` from `engine_traits`. After that I think it will be possible to delete `Peekable for Snapshot` from `engine` and migrate all Snapshot Peekable usage to `engine_rocks`.


###  What is the type of the changes?

Pick one of the following and delete the others:

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

cargo test --all --no-run

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

no

###  Does this PR affect `tidb-ansible`?

no

###  Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/4184

###  Benchmark result if necessary (optional)

###  Any examples? (optional)

zhangjinpeng1987 commented on components/engine_traits/src/options.rs:
```diff
@@ -147,8 +171,8 @@ impl Default for IterOptions {
-            fill_cache: false,
-            key_only: false,
+            fill_cache: true,
+            titan_key_only: false,
```
I think `key_only` does not equal to `titan_key_only`, when we don't use titan we also can use key_only, for example select count(*) doesn't read values but only read committed information in write cf.

Connor1996 replied:
@zhangjinpeng1987 As you said, `select count(*) doesn't read values` is controlled by `omit_value` on scanner level. But here `key_only` is actually for engine level. Cause `Rocksdb` doesn't support key_only and has no related options, the `key_only` is now for titan specifically.

Connor1996 replied:
I think we can keep the name `key_only` for other engines

zhangjinpeng1987 commented on components/engine_rocks/src/engine.rs:
```diff
@@ -105,17 +106,24 @@ impl Iterable for RocksEngine {
+    fn get_value_cf_opt(
+        &self,
+        opts: &ReadOptions,
+        cf: &str,
```
Better to move `cf` ahead of opts? Keep consistent with `iterator_cf_opt`?

Little-Wallace replied:
+1

brson replied:
There are a bunch of these situations that need to be consistent and putting cf in front of opts would be consistent with the `Mutable` trait, which currently looks like this:

```
pub trait Mutable {
    fn put_opt(&self, opts: &WriteOptions, key: &[u8], value: &[u8]) -> Result<()>;
    fn put_cf_opt(&self, opts: &WriteOptions, cf: &str, key: &[u8], value: &[u8]) -> Result<()>;

    fn delete_opt(&self, opts: &WriteOptions, key: &[u8]) -> Result<()>;
    fn delete_cf_opt(&self, opts: &WriteOptions, cf: &str, key: &[u8]) -> Result<()>;

    fn put(&self, key: &[u8], value: &[u8]) -> Result<()> {
        self.put_opt(&WriteOptions::default(), key, value)
    }

    fn put_cf(&self, cf: &str, key: &[u8], value: &[u8]) -> Result<()> {
        self.put_cf_opt(&WriteOptions::default(), cf, key, value)
    }

    fn delete(&self, key: &[u8]) -> Result<()> {
        self.delete_opt(&WriteOptions::default(), key)
    }

    fn delete_cf(&self, cf: &str, key: &[u8]) -> Result<()> {
        self.delete_cf_opt(&WriteOptions::default(), cf, key)
    }
    ...
}

When the `_opt` functions are viewed as refinements of the non-opt functions, shuffling the `cf` argument around to be in front of the introduced `opts` argument is inconsistent: `cf` and `key` always go together, so they shouldn't be split up by options.

If you want to rethink the order of arguments to these functions I would suggest either putting `opts` in front of all arguments or behind all arguments, but not in the middle.

I would be reluctant to change other functions not involved in this patchset though.

zhangjinpeng1987 commented on components/engine_traits/src/options.rs:
```diff
@@ -43,13 +41,19 @@ impl Default for WriteOptions {
     fill_cache: bool,
-    key_only: bool,
+    // only supported when Titan enabled, otherwise it doesn't take effect.
+    titan_key_only: bool,
```
@Connor1996 PTAL

zhangjinpeng1987 commented on src/raftstore/store/region_snapshot.rs:
```diff
@@ -258,7 +264,10 @@ impl RegionIterator {
         let start_key = iter_opt.lower_bound().unwrap().to_vec();
         let end_key = iter_opt.upper_bound().unwrap().to_vec();
-        let iter = snap.db_iterator_cf(cf, iter_opt).unwrap();
+        let iter = snap
```
@BusyJay PTAL

zhangjinpeng1987 commented:
rest LGTM

Connor1996 commented on components/engine_rocks/src/engine.rs:
```diff
@@ -105,17 +106,24 @@ impl Iterable for RocksEngine {
         let handle = get_cf_handle(&self.0, cf)?;
         let v = self.0.get_cf_opt(handle, key, &opt.into_raw())?;
-        Ok(v.map(|v| v.to_vec()))
+        Ok(v.map(RocksDBVector::from_raw))
```
can we return `self.0.get_cf_opt(handle, key, &opt.into_raw())` directly?

brson replied:
No, because RocksDBVector is not the same as the DBVector from the rocks crate and can't trivially be made to be the same type.

Connor1996 commented on components/engine_rocks/src/db_vector.rs:
```diff
@@ -0,0 +1,36 @@
+
+pub struct RocksDBVector(RawDBVector);
+
+impl RocksDBVector {
```
I'm wondering why we need to wrap `RawDBVector` with `RocksDBVector`? why not use `rocksdb::DBVector` directly?

brson replied:
Without wrapping it it can't implement the DBVector trait.

Connor1996 replied:
but `RawDBVector` does implement Debug + Deref<Target = [u8]> + for<'a> PartialEq<&'a [u8]> {}

brson replied:
Yes. It could be made to automatically implement DBVector with a blanket implementation. It's not obvious to me that that is a better solution than the newtype written here. The implementation here has the advantage that it follows a pattern that is consistent with the rest of the engine_rocks crate: defining the types that implement the traits from engine_traits. Introducing a new pattern so that an out-of-crate type can implement an engine trait increases the cognitive burden of maintaining the crate.

Connor1996 replied:
got it

brson commented:
This is updated for master and https://github.com/tikv/tikv/pull/5835, changes `titan_key_only` to `key_only` and addresses comments. It does not reorder the arguments of `get_value_cf` per my comments.

brson commented:
/test

brson commented:
/test

brson commented:
Several test failures. Seems like they might be real. I'll try to reproduce them.

brson commented:
/test

brson commented:
I tagged this with DNM since the CI logs keep showing core dumps.

brson commented:
/test

brson commented:
/test
